[0.1.0] (ZH-12) 기본 설정 파일 및 필수 Bean 등록

son-daehyeon opened the pull request:
### 📌 Task Details
- [x] application.yaml 분리 및 작성
- [x] MongoDB 설정 (Base Schema 설정)
- [x] Redis 설정

---

### 💬 Review Requirements (Optional)

son-daehyeon committed:
[ZH-12] chore(git): .env 파일 gitignore에 추가

son-daehyeon committed:
[ZH-12] chore(yml): application.yaml 작성

son-daehyeon committed:
[ZH-12] feat(redis): Redis 설정

son-daehyeon committed:
[ZH-12] feat(mongodb): MongoDB 설정

son-daehyeon committed:
[ZH-12] chore(util): EnableSpringDataWebSupport 설정

son-daehyeon committed:
[ZH-12] feat(schema): 슈퍼 스키마 구현

EunjinWoo commented on src/main/java/com/github/zighang_zighang/global/property/RedisProperty.java:
```diff
@@ -0,0 +1,34 @@
+    @NotBlank
+    String host;
+
+    @Min(1)
```
옹 `@Validated` 없이도 이런 validation 어노테이션 잘 동작하나요 ?!

son-daehyeon replied:
깜빡했습니다... 수정하겠습니다

EunjinWoo commented on src/main/java/com/github/zighang_zighang/global/config/RedisConfig.java:
```diff
@@ -0,0 +1,43 @@
+        redisStandaloneConfiguration.setHostName(redisProperty.getHost());
+        redisStandaloneConfiguration.setPort(redisProperty.getPort());
+        redisStandaloneConfiguration.setUsername(redisProperty.getUsername());
+        redisStandaloneConfiguration.setPassword(redisProperty.getPassword());
```
프로퍼티 보니까 password Nullable인데 NPE 방지하는 로직 없어도 되나요,,?! 잘 몰라셔..

son-daehyeon replied:
오히려 의도했던 바 입니다!

서버에서 레디스 비밀번호를 쓰지 않는 경우는 `redisConfiguration.setPassword(null)` 로 명시적으로 설정하여 비밀번호를 생략하고자 하였습니다.

`redisProperty` 자체는 `null`일 수 없으니 위 로직에서는 NPE가 발생할 원인이 없다고 생각했어요.

(혹시 다른 의견이시면 다시 코멘트 달아주세요)

EunjinWoo replied:
오오오 그렇군요! 

제가 알기로는 `redisStandaloneConfiguration.setPassword(null);`
이 들어갔을 때 내부적으로 Spring Data Redis의 `RedisStandaloneConfiguration.setPassword(String)`가 내부에서 password.toCharArray() 같은 처리를 하거나 notNull로 막아서 null을 넘기면 바로 예외가 나는 걸로 알고 있었는데 저두 더 알아봐야겠네요 .. 🤔

son-daehyeon replied:
방금 추적해봤는데,

### 1. RedisConfiguration
```java
default void setPassword(@Nullable String password) {
    setPassword(RedisPassword.of(password));
}
```

### 2. RedisPassword
```java
public static RedisPassword of(@Nullable String passwordAsString) {

    return Optional.ofNullable(passwordAsString)
            .filter(StringUtils::hasText)
            .map(it -> new RedisPassword(it.toCharArray()))
            .orElseGet(RedisPassword::none);
}
```

이렇게 옵셔널 체이닝으로 쓰고있어서 안전해요!

EunjinWoo commented on src/main/java/com/github/zighang_zighang/global/property/RedisProperty.java:
```diff
@@ -0,0 +1,34 @@
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.context.annotation.Configuration;
+
+@Data
```
`@Data` 대신 `@Getter` 정도만 쓰는 건 어떤가요?! 데이터 어노테이션 쓴 이유가 궁금합니당

son-daehyeon replied:
제가 롬복을 남용하는 편이라서 ㅎㅎ...

가끔 프로퍼티가 잘 매핑되었나 확인할 때 로그찍어보는데, `@ToString` 붙여야 하기도 귀찮아서 `@Data` 넣어놨어요.

지금 봤는데, 필드가 불변이 아니라 setter로 수정할 수 있겠네요

팀 컨벤션 정하면 좋을 것 같은데 최소한의 롬복을 쓰고자 하면 `@Getter`랑 `@ToString`으로 수정하여야 할 것 같아요.
아니라면 필드 자체를 불변으로 수정하겠습니다!

EunjinWoo replied:
아아 좋아요 ㅎㅎ 저도 롬복에 대해 아직 잘 알고 쓰진 못하는 편이라,, 같이 컨벤션 맞춰보면 좋을 것 같아요! 최소한으로 쓰는게 좋다고는 알고 있는데 이렇게 쓰는게 더 편할 거 같긴해서 대현꿍의 의견에 따르겠습니당

son-daehyeon replied:
으악 아니에요 저도 최소한으로 쓰는게 좋은거로 알고있어서 게터랑 투스트링만 달아뒀어요

EunjinWoo reviewed (approved):
고생했어요 대현꿍 빠른 작업 고마와요

son-daehyeon committed:
[ZH-12] fix(property): `@Data` 대신 `@Getter` 및 @ToString` 사용

https://github.com/zighang-zighang/zighang-zighang-backend/pull/5#discussion_r2280890599

son-daehyeon committed:
[ZH-12] fix(property): 검증 어노테이션 추가

https://github.com/zighang-zighang/zighang-zighang-backend/pull/5#discussion_r2280887807

EunjinWoo reviewed (approved):
좋습니당 !! 👍👍
